File: client/src/frame.cpp
#include "frame.hpp"
#include "opcode.hpp"
#include <array>
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>
#include <winsock2.h>

Frame::Frame(
  bool fin,
  Opcode opcode,
  bool masked,
  int payloadLength,
  std::array<byte, 4> maskingKey,
  std::vector<char> payload
) {
  this->fin = fin;
  this->opcode = opcode;
  this->masked = masked;
  this->payloadLength = payloadLength;
  this->maskingKey = maskingKey;
  this->payload = payload;
}

int Frame::recvData(SOCKET *s, char* buf, int len) {
  int result = recv(*s, buf, len, 0);
  if (result == SOCKET_ERROR) {
    logger.errorf("recv returned with error {}", WSAGetLastError());
    throw new std::runtime_error("recv returned with an error");
  }
  return result;
}

bool Frame::validMaskingKey(std::array<byte, 4> key) {
  if (
    key[0] == 0 &&
    key[1] == 0 &&
    key[2] == 0 &&
    key[3] == 0
  ) return false;
  return true;
}

std::array<byte, 4> Frame::getMaskingKey() {
  std::array<byte, 4> out;
  for (int i = 0; i < 4; i++) {
    out[i] = std::rand() % 255 + 1;
  }
  return out;
}

const std::string chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
std::string Frame::getConnectionKey() {
  std::string out;
  for (int i = 0; i < 16; i++) {
    out += chars[std::rand() % chars.length()];
  }
  return out;
}

std::string Frame::getPayloadString() {
  return std::string(payload.begin(), payload.end());
}

Frame Frame::read(SOCKET* s) {
  if (s == nullptr || *s == INVALID_SOCKET) {
    logger.error("An invalid socket has been provied to Frame.read()");
    throw new std::invalid_argument("Provided socket is invalid");
  }
  char info_buf[2];
  int result = recvData(s, info_buf, sizeof(info_buf));

  char b = info_buf[0];
  boolean fin = (b & 0x80) != 0;
  Opcode opcode = toOpcode(b & 0x0F);
  b = info_buf[1];
  boolean masked = (b & 0x80) != 0;
  int payloadLength = b & 0x7F;
  int bytesToRead = 0;

  if (payloadLength == 126) bytesToRead = 2;
  if (payloadLength == 127) bytesToRead = 8;
  if (bytesToRead != 0) {
    char payload_len_buf[bytesToRead];
    recvData(s, payload_len_buf, strlen(payload_len_buf));
    payloadLength = 0;
    for (int i = 0; i < bytesToRead; i++) {
      payloadLength = (payloadLength << 8) + (payload_len_buf[i] & 0xFF);
    }
  }

  std::array<byte, 4> maskingKey;
  if (masked) {
    char mask_buf[4];
    recvData(s, mask_buf, 4);
    for (int i = 0; i < 4; i++) {
      maskingKey[i] = mask_buf[i];
    }
  }
  
  char payload_buf[payloadLength];
  recvData(s, payload_buf, payloadLength);
  std::size_t size = sizeof(payload_buf) / sizeof(payload_buf[0]);
  std::vector<char> payload(payload_buf, payload_buf + size);
  if (masked) {
    for (int i = 0; i < payloadLength; i++) {
      payload[i] = payload[i] ^ maskingKey[i % 4];
    }
  }

  return Frame(fin, opcode, masked, payloadLength, maskingKey, payload);
}

std::vector<char> Frame::getBytes() {
  if (payloadLength != payload.size()) {
    logger.error("payload.size() does not match payloadLength when trying to call getBytes()");
    throw new std::invalid_argument("payload.size() does not match payloadLength when trying to call getBytes()");
  }
  int calcLength = 2 + payloadLength; // 2 = min len;
  if (payloadLength <= 65535 && payloadLength > 125) calcLength += 3; // 1 byte + next 2
  else if (payloadLength > 65535) calcLength += 9; // 1 byte + next 8

  if (masked && !validMaskingKey(maskingKey)) {
    logger.error("Masked set to true but frame does not contain valid maskingKey");
    throw new std::invalid_argument("Masked set to true but frame does not contain valid maskingKey");
  } else if (masked) {
    calcLength += 4;
  }

  std::vector<char> out;
  out.resize(calcLength);
  int idx = 0;

  if (fin) out[idx] |= 0b10000000;
  out[idx] |= static_cast<byte>(opcode);
  idx++;

  if (payloadLength <= 125) {
    if (masked) out[idx] |= 0b10000000;
    out[idx] |= payloadLength;
    idx += 1;
  } else if (payloadLength <= 65535){
    if (masked) out[idx] = (byte)(0b10000000);
    out[idx] |= 126;
    out[idx + 1] = (byte)((payloadLength >> 8) & 0xFF); 
    out[idx + 2] = (byte)(payloadLength & 0xFF);
    idx += 3;
  } else {
    if (masked) out[idx] = (byte)(0b10000000);
    out[idx] |= 127;
    idx += 1;
    for (int i = 7; i >= 0; i--) {
      out[idx + 8 - i] = (byte)((payloadLength >> (i * 8)) & 0xFF);
      idx += 1;
    }
  }

  if (masked) {
    for (int i = 0; i < 4; i++) {
      out[idx] = maskingKey[i];
      idx++;
    }
    for (int i = 0; i < payloadLength; i++) {
      out[idx] = payload[i] ^ maskingKey[i % 4];
      idx++;
    }
  } else {
    out.insert(out.begin() + idx, payload.begin(), payload.end());
  }
  return out;
}

Frame Frame::getGeneric(Opcode opcode, bool masked) {
  std::array<byte, 4> maskingKey;
  if (masked) maskingKey = getMaskingKey();
  return Frame(true, opcode, masked, 0, maskingKey, std::vector<char>());
}

Frame Frame::getClose(std::string reason, bool masked) {
  std::vector<char> payload(reason.begin(), reason.end());
  std::array<byte, 4> maskingKey;
  if (masked) maskingKey = getMaskingKey();
  return Frame(true, Opcode::CLOSE, masked, payload.size(), maskingKey, payload);
}

File: controller/src/main/java/io/github/oxi1224/remotecontroller/server/Frame.java
package io.github.oxi1224.remotecontroller.server;

import java.io.IOException;
import java.io.InputStream;

public class Frame {
  public final boolean fin;
  public final Opcode opcode;
  public final boolean masked;
  public final int payloadLength;
  public final byte[] maskingKey;
  public final byte[] payload;

  public Frame(
    boolean fin,
    Opcode opcode,
    boolean masked,
    int payloadLength,
    byte[] maskingKey,
    byte[] payload
  ) {
    this.fin = fin;
    this.opcode = opcode;
    this.masked = masked;
    this.payloadLength = payloadLength;
    this.maskingKey = maskingKey == null ? new byte[0] : maskingKey;
    this.payload = payload == null ? new byte[0] : payload;
  }

  public byte[] getBytes() {
    int calcLength = 2 + payloadLength; // 2 = min len;
    if (payloadLength <= 65535 && payloadLength > 125) calcLength += 3; // 1 byte + next 2
    else if (payloadLength > 65535) calcLength += 9; // 1 byte + next 8
    if (masked && maskingKey.length != 4) {
      throw new IllegalArgumentException("The masking key must be an array of 4 bytes");
    } else if (masked ) {
      calcLength += 4;
    }
    byte[] out = new byte[calcLength];
    int idx = 0;
    byte b = 0x0;
    if (fin) b |= 0b10000000;
    b |= opcode.getValue();
    out[idx] = (byte)(b & 0xFF);
    idx += 1;

    if (payloadLength <= 125) {
      b = 0x0;
      if (masked) b |= 0b10000000;
      b |= payloadLength;
      out[idx] = b;
      idx += 1;
    } else if (payloadLength <= 65535){
      if (masked) out[idx] = (byte)(0b10000000);
      out[idx] |= 126;
      out[idx + 1] = (byte)((payloadLength >> 8) & 0xFF); 
      out[idx + 2] = (byte)(payloadLength & 0xFF);
      idx += 3;
    } else {
      if (masked) out[idx] = (byte)(0b10000000);
      out[idx] |= 127;
      idx += 1;
      for (int i = 7; i >= 0; i--) {
        out[idx + 8 - i] = (byte)((payloadLength >> (i * 8)) & 0xFF);
        idx += 1;
      }
    }
    if (masked) {
      for (int i = 0; i < 4; i++) {
        out[idx] = maskingKey[i];
        idx++;
      }
      /// TODO: masking
    } else {
      System.arraycopy(payload, 0, out, idx, payloadLength);
    }
    return out;
  }

  public static Frame read(InputStream in) throws IOException {
    byte b = (byte)(in.read()); 
    boolean fin = (b & 0x80) != 0;
    Opcode opcode = Opcode.findByValue(b & 0x0F);

    b = (byte)(in.read());
    boolean masked = (0x80 & b) != 0;
    int payloadLength = (byte)(0x7F & b);
    int bytesToRead = 0;
    if (payloadLength == 126) bytesToRead = 2;
    if (payloadLength == 127) bytesToRead = 8;
    if (bytesToRead != 0) {
      byte[] bytes = in.readNBytes(bytesToRead);
      payloadLength = 0; 
      for (byte _b : bytes) {
        payloadLength = (payloadLength << 8) + (_b & 0xFF);
      }
    };
    byte[] maskingKey = null;
    if (masked) {
      maskingKey = new byte[4];
      in.read(maskingKey, 0, 4);
    }
    byte[] payload = new byte[payloadLength];
    in.read(payload, 0, payloadLength);
    return new Frame(fin, opcode, masked, payloadLength, maskingKey, payload);
  }

  public static byte[] getMaskingKey() {
    byte[] key = new byte[4];
    for (int i = 0; i < 4; i++) {
      key[i] = (byte)(Math.random() * 255);
    }
    return key;
  }

  public static Frame getGeneric(Opcode opcode, boolean masked) {
    byte[] key = masked ? getMaskingKey() : null;
    return new Frame(true, opcode, masked, 0, key, null);
  }

  public static Frame getClose(String reason, boolean masked) {
    byte[] key = masked ? getMaskingKey() : null;
    byte[] payload = reason.getBytes();
    return new Frame(true, Opcode.CLOSE, masked, payload.length, key, payload);
  }
}
